hudi-flink-datasource/hudi-flink/src/test/java/org/apache/hudi/sink/utils/TestWriteBase.java

```diff
@@ -524,9 +524,15 @@ private void checkInstantState(HoodieInstant.State state, String instantStr) {
     }
 
     protected String lastCompleteInstant() {
-      return OptionsResolver.isMorTable(conf)
-          ? TestUtils.getLastDeltaCompleteInstant(basePath)
-          : TestUtils.getLastCompleteInstant(basePath, HoodieTimeline.COMMIT_ACTION);
+      if (OptionsResolver.isLocklessMultiWriter(this.conf)) {
+        // If there are lockless multiple writer, fetch last complete instant of current writer from ckp metadata
+        // because the writer started first might finish later, similarly, the writer started later might finish first.
+        return this.ckpMetadata.lastCompleteInstant();
+      } else {
```

Comment:
Got it, maybe we can use `return this.ckpMetadata.lastCompleteInstant();` directly.

Reply:
In most case `return this.ckpMetadata.lastCompleteInstant();`  directly works fine.
But the case `testSubtaskFails` would fail because failover happens, the ckp meta might be cleaned, fetch the instant from timeline.
I add branch to handle this case.

Reply:
Does `this.ckpMetadata.isEmpty()` indicate a failover?